src/System.Text.Http.Parser/HttpParser.cs

```diff
@@ -340,6 +340,94 @@ public class HttpParser : IHttpParser
+            int nameStart = 0, nameEnd = 0, valueStart = 0, valueEnd = 0;
+            bool sawCRLF = false;
+            bool straddling = false;
+            fixed (byte* pBuffer = &span.DangerousGetPinnableReference()) {
+                while (true) {
+                    index++;
+                    // get next byte 
+                    while (index == length) {
+                        if (rest == null) {
+                            RejectRequest(RequestRejectionReason.InvalidRequestHeadersNoCRLF);
+                        }
+                        span = rest.First.Span;
+                        rest = rest.Rest;
+                        index = -1;
+                        straddling = true;
+                        throw new NotImplementedException("call to OnHeader below needs to account for straddling spans");
+                    }
+                    var next = pBuffer[index];
+
+                    if (state == State.ParsingValue) {
+                        if (next == ByteCR) {
+                            valueEnd = index;
+                            state = State.ArferCR;
+                        }
+                        // TODO: should value characters be validated?
+                        // TODO: should whitespace after all values (i.e. right before CRLF) be removed here?
+                        // else just keep advancing chracters in name
+                    }
+                    else if (state == State.ParsingName) {
+                        if (next == ByteColon) {
+                            state = State.BeforValue;
+                            nameEnd = index;
+                        }
+                        if (next == ByteCR) state = State.ArferCR;
+                        // TODO: should name characters be validated?
+                        // else just keep advancing chracters in name
+                    }
+                    else if (state == State.BeforValue) {
+                        if (next != ByteSpace && next != ByteTab) {
+                            state = State.ParsingValue;
+                            valueStart = index;
+                        }
+                        else if (next == ByteCR) RejectRequest(RequestRejectionReason.InvalidCharactersInHeaderName);
+                        // else skip whitespace
+                    }
+                    else if (state == State.ArferCR) {
+                        if (next == ByteLF) {
+                            sawCRLF = true;
+                            if (nameEnd != 0) {
+                                if (!straddling) {
+                                    ReadOnlySpan<byte> nameBuffer = new ReadOnlySpan<byte>(pBuffer + nameStart, nameEnd - nameStart);
+                                    ReadOnlySpan<byte> valueBuffer = new ReadOnlySpan<byte>(pBuffer + valueStart, valueEnd - valueStart);
+                                    handler.OnHeader(nameBuffer, valueBuffer);
+                                    nameEnd = 0;
```

Comment:
I use it in state == State.ArferCR to discover empty line, in which case nameEnd will never be set.